clients/banking/src/utils/popup.ts

```diff
@@ -0,0 +1,79 @@
+import { Option, Result } from "@swan-io/boxed";
+import { isNotNullish } from "@swan-io/lake/src/utils/nullish";
+import { isMatching } from "ts-pattern";
+
+const isClosePopupData = isMatching({
+  source: "banking",
+  event: "closePopup",
+});
+
+type GuardType<T> = T extends (value: unknown) => value is infer R ? R : never;
+type ClosePopupData = GuardType<typeof isClosePopupData>;
+
+let callback: (() => void) | undefined = undefined;
+
+const onMessage = ({ data, origin }: { data: unknown; origin: string }) => {
+  if (
+    origin === window.location.origin && // prevent cross-origin issues
+    isClosePopupData(data) &&
+    isNotNullish(callback)
+  ) {
+    window.removeEventListener("message", onMessage);
+    callback();
+    callback = undefined;
+  }
+};
+
+export const openPopup = ({ url, onClose }: { url: string; onClose: typeof callback }) => {
+  const height = 800;
+  const width = 500;
+
+  const screenLeft = window.screenLeft ?? window.screenX;
+  const screenRight = window.screenTop ?? window.screenY;
+  const { outerHeight, outerWidth } = window;
+
+  const params = [
+    `left=${screenLeft + outerWidth - (width / 2 + outerWidth / 2)}`,
+    `top=${screenRight + outerHeight - (height / 2 + outerHeight / 2)}`,
+    `height=${height}`,
+    `width=${width}`,
+    "menubar=no",
+    "status=no",
+    "toolbar=no",
+  ];
+
+  const popup = window.open(url, "Swan", params.join(","));
+
+  if (isNotNullish(popup)) {
+    // listen for messages from the popup window
+    window.addEventListener("message", onMessage);
+
+    callback = () => {
+      onClose?.(); // trigger user callback
+      popup?.close?.();
+    };
+
+    // https://stackoverflow.com/a/48240128
+    const interval = setInterval(() => {
+      if (popup?.closed === true) {
+        window.removeEventListener("message", onMessage);
+        callback = undefined;
+        clearInterval(interval);
+      }
+    }, 100);
```

Comment:
the interval clear for the nominal case seems too indirect to me here, it relies on:

- message received
- popup closes successfully
- next interval tick sees that `closed` is `true`
- GC can cleanup `popup`

I'd either go for the message listener to be initialized within the popup opening and clearing the interval or keep the `interval` in a higher scope.